### PR TITLE
Add support for building to ARM64

### PR DESCRIPTION
**Pull request by pmsjt**
Produced ARM64 driver tested on Samsung Galaxy Book2.

### CODE REVIEW

**Comment by fyodor** [comment]
Thanks Pedro.  We are looking into this and I also just added issue #1590 related to ARM64 support.

**Comment by pmsjt** [comment]
Hi folks. Is there anything else (other than the conflicts) in the way of getting this merged to baseline?

Thanks.

**Comment by fyodor** [comment]
@pmsjt - This is a very exciting improvement, and your patch helps immensely toward the "proof of concept" and implementation of ARM64 support.  But supporting a new platform will still be a lot of work.  The main Npcap developers will need to obtain ARM64 hardware, add it to the build process, test it, etc.  And there is a lot of pressing Npcap work right now (you can see our CHANGELOG at https://npcap.org/changelog to see what we've been up to most recently).  So it's mostly waiting for us to have the time to commit to this project, including buying hardware and so forth.  It also depends on how many Npcap users and customers are asking for this support (not a lot so far, but this is a fairly new Windows architecture and its popularity may keep increasing).

**Comment by guyharris** [comment]
This change needs rebasing.

**Comment by pmsjt** [comment]
> This change needs rebasing.

Rephrasing my question from above:

Is there anything else (other than the rebase) in the way of getting this merged?

Thanks.

File: packetWin7/NPFInstall/NPFInstall/NPFInstall.cpp [hunk 1]
```diff
@@ -72,39 +72,39 @@ BOOL PacketInstallDriver40();
 BOOL PacketStopDriver40();
 
 #define STR_COMMAND_USAGE \
-_T("NPFInstall for Npcap ") _T(WINPCAP_VER_STRING) _T(" ( http://npcap.org )\n")\
```
Comment (guyharris) [inline packetWin7/NPFInstall/NPFInstall/NPFInstall.cpp:75]: Does this change do anything other than add a space before the backslash escape at the end of the line?

If so, is it necessary to support newer versions of MSVC?

File: packetWin7/WlanHelper/WlanHelper/WlanHelper.cpp [hunk 2]
```diff
@@ -517,48 +517,48 @@ BOOL SetWlanOperationMode(tstring strGUID, tstring strMode)
 }
 
 #define STR_COMMAND_USAGE \
-_T("WlanHelper for Npcap ") _T(WINPCAP_VER_STRING) _T(" ( http://npcap.org )\n")\
```
Comment (guyharris) [inline packetWin7/WlanHelper/WlanHelper/WlanHelper.cpp:520]: Does this change do anything other than add a space before the backslash escape at the end of the line?

If so, is it necessary to support newer versions of MSVC?

File: packetWin7/npf/npf/Packet.c [hunk 3]
```diff
@@ -196,10 +196,12 @@ My_NdisGroupMaxProcessorCount(
 			Cpu = NPF_MAX_CPU_NUMBER - 1;
 		}
 	}
+#if NDIS_MIN_API < 0x630
 	else // for NDIS6 (Vista)
```
Comment (guyharris) [inline packetWin7/npf/npf/Packet.c:200]: Vista is no longer supported; does ARM64 Windows not support NDIS 6.3?

File: packetWin7/npf/npf/Read.c [hunk 4]
```diff
@@ -1026,13 +1026,13 @@ NPF_TapExForEachOpen(
 						{
 							//the Radiotap header will be fragmented in the buffer (aka, it will skip the buffer boundary)
 							ToCopy = Open->Size - LocalData->P;
-							NdisMoveMappedMemory(LocalData->Buffer + LocalData->P, Dot11RadiotapHeader, ToCopy);
-							NdisMoveMappedMemory(LocalData->Buffer + 0, Dot11RadiotapHeader + ToCopy, Dot11RadiotapHeaderSize - ToCopy);
+							RtlCopyMemory(LocalData->Buffer + LocalData->P, Dot11RadiotapHeader, ToCopy);
```
Comment (guyharris) [inline packetWin7/npf/npf/Read.c:1029]: This appears already to have been done in the tip of the master branch, so it should no longer be necessary (`NdisMoveMappedMemory()` is no longer being called).

**Comment by guyharris** [comment]
> > This change needs rebasing.
> 
> Rephrasing my question from above:
> 
> Is there anything else (other than the rebase) in the way of getting this merged?

There are several parts to this change:

- updating the project files to a newer version of Visual Studio, which is also issue #197 - the developers would probably need to decide to make that change, and make sure it doesn't break anything, so it would require some attention on their part, and, if they make that change, that would mean that the changes in this pull request would no longer be necessary;

- the changes that add a space before the backslash escape at the end of lines in a couple of source files, which, if they're not a requirement for ARM64 support, should probably be removed from this pull request and put into a separate pull request, to simplify the process of reviewing this pull request;

- the changes to the project files to add ARM64 as a platform;

- the changes to the code to use `_WIN64` rather than `_AMD64_` as the test for "is this 64-bit?"

- the change to packetWin7/npf/npf/Packet.c not to call `NdisSystemProcessorCount()` if `NDIS_MIN_API < 0x630`, which no longer applies, as Vista support was dropped;

- the change to packetWin7/npf/npf/Read.c to use `RtlCopyMemory()` rather than `NdisMoveMappedMemory()`, which no longer applies, as it appears already to have been done.

So:

- you will probably have to wait for the project to decide to upgrade the build to use VS 2017 or later, and fix #197, and then rebase (or start from the tip of the master branch and reapply those changes that are still necessary).

- after that, the only changes appear to be adding NTarm64 to packetWin7/npf/npf/npf.inf and packetWin7/npf/npf/npf_wfp.inf, any other project file changes that would remain after the rebasing, and  changing the code to use `_WIN64` rather than `_AMD64_` as the test for "is this 64-bit?".

At that point, they might be willing to merge the changes, after testing to make sure that they don't break 32-bit x86, 64-bit x86, or IA-64 (if that's supported) builds, but that wouldn't necessarily mean that those changes would be "supported" by the Npcap developers in any sense more than "if you want Npcap for ARM64, grab the code and build it - good luck!"  *That's* where the work @fyodor's talking about would have to happen, and that might take a while, for the reasons he states; the changes would be in the code base, but they wouldn't be an official part of Npcap, and they wouldn't ship ARM64 binaries, until they do all that work.

**Comment by pmsjt** [comment]
> > Rephrasing my question from above:
> > Is there anything else (other than the rebase) in the way of getting this merged?

I am glad I asked. There is, at least one thing here that is out of my control. Setting myself to rebased/resolve would just be a waste of time. When this blocker is removed, chances are we'd need yet another rebased, as time would have gone by.

> * the changes to the code to use `_WIN64` rather than `_AMD64_` as the test for "is this 64-bit?"

Which is a trivial bug in the current payload (would have not worked for Itanium either)

> * you will probably have to wait for the project to decide to upgrade the build to use VS 2017 or later, and fix #197, and then rebase (or start from the tip of the master branch and reapply those changes that are still necessary).

This is the only real blocking issue here, if I understand the comment. YS2017 is a requirement for ARM64. Given its age, I'd say it is about time, but that is, like you pointed out, up to the project maitainers to decide.

>  "if you want Npcap for ARM64, grab the code and build it - good luck!" _That's_ where the work @fyodor's talking about would have to happen, and that might take a while, for the reasons he states; the changes would be in the code base, but they wouldn't be an official part of Npcap, and they wouldn't ship ARM64 binaries, until they do all that work.

Maybe it is just me, but there is commonly an understanding of supported binary releases vs code support for other platforms. This is the case, for example, with Debian Ports "we don't officially release it, but it ought work if you build it".

So, what is the next step here @fyodor ? Here's my suggestion:
* Abandon this PR;
* Open an issue for tracking the (the decision to) upgrade the project to VS2017
* Wait for the decision to be made before opening an updated PR again.

Does this sound reasonable and a good use of everyone's time?

Thanks.

**Comment by guyharris** [comment]
> So, what is the next step here @fyodor ? Here's my suggestion:

...

> * Open an issue for tracking the (the decision to) upgrade the project to VS2017

That's already open - issue #197.

**Comment by guyharris** [comment]
OK, #197 has been closed, so you should take the tip of the master branch, make the changes again using the same version of Visual Studio that they're using (or a newer version), and submit a new pull request.

From the changes in 665ae79371e8fdad5a880a147341ade1edf06608, which was the change that closed #197, the PlatformToolset changed from v120 to v142 and the WindowsTargetPlatformVersion changed from 10.0.14393.0 to 10.0.18362.0.

**Comment by pmsjt** [comment]
Ok @guyharris 
Rebased and tested to top-of-tree.

**Comment by pmsjt** [comment]
@guyharris & @fyodor , here are the updated test binaries: 

https://1drv.ms/u/s!Ag4DqZg51HWgvs5zTpttv5Mchi6S1A?e=ikiL1q

As before, install the drivers and copy the respective npcap folders to system32 and syswow64.

If you test this on a Windows build with x86_64 compatibility, then the both x86_64 and x86_32 versions of WireShark work.

**Comment by fyodor** [comment]
Thanks @pmsjt!

**Comment by bader29** [comment]
Xx

**Comment by pmsjt** [comment]
@fyodor conflicts have developed again. It has been a month since I made the last one but the PR was never approved nor any suggestions were made. I can rebase it yet again but I'd like to have some assurance that we won't be doing this perpetually.

**Comment by fyodor** [comment]
Hey @pmsjt .  Thanks for all of your work on this.  We would love to work more on the ARM port, but we have spent much of the last month dealing with MIcrosoft driver signature verification and WHQL issues on the x86 and x64 platforms.  Sorry to go a little off topic on this pull request, but If you (or anyone reading this) knows anyone (at Microsoft or otherwise) who might be able to help (for free or paid), please let us know.  Here are some details that @dmiller-nmap sent me:

> Since Windows 10 1607, Microsoft has required drivers to be WHQL-certified or MS Attestation-signed before they can be installed. We are currently pursuing WHQL certification, but for existing releases we used the attestation-signing process. We test our driver, sign it with our EV code-signing certificate, and submit it to Microsoft with our attestation that we have tested it for these platforms. They perform cursory verification of the drivers we submit and then add an additional signature. This is the "Microsoft Windows Hardware Compatibility Publisher" certificate, which should already be trusted on those systems.
>
> For all other (non Windows 10) platforms, we use our SHA-256 EV code signing certificate. We recently renewed this prior to the 5/7/2021 expiration, but we have not made a release with the new certificate yet. On those platforms, we install (via certutil) our own certificate prior to attempting installation. The driver also contains a cross-certificate for kernel-mode driver code signing issued by Microsoft which further certifies the authority of our CA to issue certificates valid for kernel-mode driver code signing. Without this cross-certificate, Windows will not allow the driver to be installed.
>
> For Windows 7, we have taken several approaches to address the possibility of an unpatched system which cannot validate a SHA-256 signature. For Npcap releases prior to 0.9985, the driver was dual-signed with matching SHA-256 and SHA-1 certificates. For 0.9985 through 0.9990, we used a different CA who only issued SHA-256 certificates, so these releases are incompatible with unpatched Windows 7 systems. Npcap 0.9991 through 0.9997 reverted to the dual-signed drivers, which still were causing problems for some users. In Npcap 1.00 and later, we chose to only use the SHA-256 certificate for all platforms, but we also include a SHA-1-only signed version for Windows 7. On that platform only, if the SHA-256-signed driver fails to install, we will attempt to install the SHA-1-signed driver instead.
>
> For future releases, we are still working out a way ahead. Microsoft has removed the cross-certificate path for driver signatures, so we are currently looking into WHQL certification. This is a much more extensive process, requiring full functional testing on every supported OS version and architecture, so we are likely to reduce the frequency of releases to accommodate the additional cost of testing. Our understanding is that when this certification is done we will ship a driver that is signed by Microsoft directly, which will finally resolve any signature validation problems.

So basically the driver signing method we were using for pre-Win10 releases no longer works because Microsoft [removed the cross-certificate path for driver signatures](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/deprecation-of-software-publisher-certificates-and-commercial-release-certificates).  Interestingly, attestation signing still works for newer (Win10+) platforms so we can still sign drivers for modern Windows releases.  But attestation signing doesn't work for earlier releases like Windows Servier 2012r2 and Windows 8, and the crocc-certificate path removal eliminated our alternative approach.  The only thing we have found which we think might still work for older Windows releases is full-on WHQL testing, which is an enormously costly and comnplex process for a small organization like ours.  We were hoping we could avoid setting up dozens of systems by running each platform test on a different VM, but that has not worked for us so far.  We have also investigated third-party certification labs, but the quotes we've received so far are in the tens of thousands of dollars (plus several weeks of testing time) per release.  And we were making roughly monthly Npcap releases until this whole situation came up, so WHQL would add a whole lot of friction.  I'm hoping there is still some way to sign pre-Win10 drivers that we're missing.

Anyway, this is a long way of saying we would love to work on the ARM port but we can't even start that until we solve this difficult and pressing signature/validation/WHQL issue.  After that we can look at cool new features to potentially work on next (like ARM support, an MSI installer, etc.)  If this rebasing is a pain, it might be easier for you to wait until we are able to actually start working on the ARM64 project, which will involve buying ARM64 hardware to test with, etc.  I hope this helps!

**Comment by pmsjt** [comment]
Hi @fyodor . 

Thanks for your response and context.

Let me see if I can recruit some help.

**Comment by fyodor** [comment]
Thanks @pmsjt!

**Comment by dmiller-nmap** [comment]
Thank you all for your help navigating this change. We have added ARM64 support in Npcap 1.40.

**Comment by fyodor** [comment]
Yeah thanks again to everyone, especially @pmsjt and of course @guyharris!  There was a major bug in the 1.40 installer and so we've re-issued the release as 1.50 this time.  We're super excited about the Windows on ARM support, so please let us know how it goes!  We purchased a Microsoft Surface Pro X tablet and also a Samsung Galaxy Book Go laptop for our lab and it works great on those!

**Pull request closed without merging**